junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvArgumentsProvider.java

```diff
@@ -54,22 +54,35 @@ public void accept(CsvSource annotation) {
 		Preconditions.condition(this.annotation.value().length > 0 ^ textBlockDeclared,
 			() -> "@CsvSource must be declared with either `value` or `textBlock` but not both");
 
-		String[] lines;
 		if (textBlockDeclared) {
-			lines = NEW_LINE_REGEX.split(this.annotation.textBlock(), 0);
-		}
-		else {
-			lines = this.annotation.value();
+			return parseTextBlock(this.annotation.textBlock()).stream().map(Arguments::of);
 		}
 
-		AtomicInteger index = new AtomicInteger(1);
+		AtomicInteger index = new AtomicInteger(0);
 		// @formatter:off
-		return Arrays.stream(lines)
-				.map(line -> parseLine(line, index.getAndIncrement()))
+		return Arrays.stream(this.annotation.value())
+				.map(line -> parseLine(line, index.incrementAndGet()))
 				.map(Arguments::of);
 		// @formatter:on
 	}
 
+	private List<String[]> parseTextBlock(String textBlock) {
+		try {
+			AtomicInteger index = new AtomicInteger(0);
+			List<String[]> csvRecords = this.csvParser.parseAll(new StringReader(stripIndent(textBlock)));
```

Comment:
Is there an API that returns an iterator or sth. like that? It's not a big concern since these strings shouldn't be huge, though.

Reply:
> Is there an API that returns an iterator or sth. like that?

There's `parse(Reader)` which delegates to a `RowProcessor`. The other option is to use `beginParsing(Reader)` and `parseNext()` (like in `CsvFileArgumentsProvider`).

So, we could use `beginParsing(new StringReader(stripIndent(textBlock))` and then iterate the same as in `CsvFileArgumentsProvider` with our `CsvParserIterator` (or investigate how to do that with a `RowProcessor`), but I think that's overkill since the entire text block string is already in memory.

> It's not a big concern since these strings shouldn't be huge, though.

Yes, that was my rationale for using `parseAll(Reader)`.